Bug/CCP-209

mkarthik90 opened the pull request.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -449,47 +455,77 @@ public void indexConcepts(String userName)
+                        query.append(" ))");
+
+                        queryString.append(query.toString());
+                        firstEntry++;
```
what about exact matches (e.g. if someone tries to find "Albert Einstein" theory)? Also, have you tried using a QueryBuilder (https://lucene.apache.org/core/5_4_0/core/org/apache/lucene/util/QueryBuilder.html?is-external=true) and BooleanQueries, etc?

jdamerow replied:
should make it much easier to read

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/reflect/LuceneField.java:
```diff
@@ -9,7 +9,17 @@
+     * 
+     * @return
+     */
+    boolean isShortWordSearchAllowed() default false;
```
this is a bad name. It's not about allowing short words, its about storing and searching for very short words.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/constants/LuceneFieldNames.java:
```diff
@@ -16,4 +16,5 @@
     public final static String WORDNETID = "wordnetid";
     public final static String CONCEPTTYPE = "conceptType";
     public final static String MERGED_IDS = "mergedIds";
+    public final static String SHORT_WORD = "short_word";
```
this is a bad field name. those field names are content specific.

mkarthik90 replied:
We can't have the same SHORT_WORD as "word" right so had as short_word.

jdamerow replied:
how about "untokenized_word"?

jdamerow replied:
it'll contain not just short words

mkarthik90 replied:
yeah sure

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
Opened to test the integration test on server. Will close once integration test completes.

mkarthik90 commented:
run integration tests please

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
run integration tests please

mkarthik90 commented:
run integration tests please

mkarthik90 commented:
run integration tests please

mkarthik90 commented:
run integration tests please

mkarthik90 commented:
@jdamerow  Addressed review comments. 

The problem was with the concept list field. The concept list field was indexed as a untokenized field. But when we are using a phrase query and if the term contains a space it doesn't work. Instead, we need to make the field tokenized and if the field contains spaces it does work. Not sure if we can make the concept list field as tokenized. Tokenization is done by changing the StringField to TextField.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -435,69 +442,66 @@ public void indexConcepts(String userName)
+                        Query q1 = shortTermsQueryBuilder.createPhraseQuery(luceneFieldAnnotation.lucenefieldName(),
+                                searchString);
+                        Query q2 = shortTermsQueryBuilder.createBooleanQuery(
+                                luceneFieldAnnotation.luceneShortFieldName(), searchString.replaceAll(" ", ""));
```
why are all blanks replaced?

jdamerow replied:
how could it find anything like that?

jdamerow replied:
how about instead of using an extra attribute of the annotation to define a field name, just make the field name: ```luceneFieldName() + "_untokenized"```? ```luceneShortFieldName``` is really not self-explanatory and just one more thing to configure

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/core/ConceptEntry.java:
```diff
@@ -51,7 +51,7 @@
 
     @SearchField(fieldName = SearchFieldNames.CONCEPT_LIST)
-    @LuceneField(lucenefieldName = LuceneFieldNames.CONCEPT_LIST, isTokenized = false, isMultiple = false)
+    @LuceneField(lucenefieldName = LuceneFieldNames.CONCEPT_LIST, isTokenized = true, isMultiple = false)
```
should not be tokenized

jdamerow replied:
needs to be an exact match. PhraseQueries can be used for that I believe:
https://lucene.apache.org/core/6_3_0/core/org/apache/lucene/search/PhraseQuery.html#PhraseQuery-java.lang.String-java.lang.String...-

mkarthik90 commented:
Opening the PR for testing the integration test results.

mkarthik90 commented:
run integration tests please

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
run integration tests please

mkarthik90 commented:
run integration tests please

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow Addressed review comments. Please review.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -192,6 +193,11 @@ public void insertConcept(ConceptEntry entry, String userName) throws LuceneExce
                     }
+
+                    if (searchFieldAnnotation.isShortPhraseSearchable()) {
+                        doc.add(new StringField(searchFieldAnnotation.lucenefieldName() + "_untokenized",
```
should be in a constant

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -293,6 +299,7 @@ private Document createIndividualDocument(IDictionary dict, IWordID wordId) {
         Document doc = new Document();
         String lemma = wordId.getLemma().replace("_", " ");
         doc.add(new TextField(LuceneFieldNames.WORD, lemma, Field.Store.YES));
+        doc.add(new StringField(LuceneFieldNames.WORD + "_untokenized", lemma, Field.Store.YES));
```
to be reused here

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -435,69 +442,67 @@ public void indexConcepts(String userName)
+                        Query q1 = shortTermsQueryBuilder.createPhraseQuery(luceneFieldAnnotation.lucenefieldName(),
+                                searchString);
+                        Query q2 = shortTermsQueryBuilder.createBooleanQuery(
+                                luceneFieldAnnotation.lucenefieldName() + "_untokenized", searchString);
```
reuse constant

jdamerow commented:
ok to test

jdamerow commented:
Searching for "be":
```
ERROR: edu.asu.conceptpower.web.util.ExceptionHandler - ExceptionHandler caught exception.
java.lang.NullPointerException: Query must not be null
	at java.util.Objects.requireNonNull(Objects.java:228)
	at org.apache.lucene.search.BooleanClause.<init>(BooleanClause.java:60)
	at org.apache.lucene.search.BooleanQuery$Builder.add(BooleanQuery.java:138)
```

jdamerow commented:
Also, search is really slow, although Lucene seems to return results very quickly. But afterwards it seems to be doing a bunch of things that slows down the search.

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
opened for integration tests

jdamerow commented:
is this ready to be reviewed again?

mkarthik90 commented:
@jdamerow   Yes please review.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/constants/LuceneFieldNames.java:
```diff
@@ -16,4 +16,5 @@
     public final static String WORDNETID = "wordnetid";
     public final static String CONCEPTTYPE = "conceptType";
     public final static String MERGED_IDS = "mergedIds";
+    public final static String UNTOKENIZED = "_untokenized";
```
let's call the constant ```UNTOKENIZED_SUFFIX``` given that it's just a suffix

jdamerow commented:
now, it doesn't seem to stem anymore ('working' doesn't find 'work'). also wildcards are not working. and it's still slow.

mkarthik90 commented:
opened for integration test results

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
run integration tests please

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
run integration tests please

mkarthik90 commented:
@jdamerow Addressed review comments. 

Major change is with integration tests. Spring was internally using XMLUnit for validating XML but there were ordering problems. So used XMLUnit explicitly in our test cases.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -435,69 +448,85 @@ public void indexConcepts(String userName)
             if (search != null) {
                 String searchString = fieldMap.get(search.fieldName());
+                if (searchString != null) {
+                    if (containsWildCardCharacters(searchString)) {
```
just add another property to the annotation to enable a field for wildcard search and then use that here.

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow  Addressed review comments.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -513,14 +508,12 @@ public void indexConcepts(String userName)
                                 shortTermBooleanQueryBuilder.add(q1, BooleanClause.Occur.SHOULD);
                             }
-
                         }
```
can't you merge the first if and the else block by always creating the shortTermsQueryBuilder, but only if it's also shortPhrase searchable create q2? Would make much cleaner code.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -484,21 +484,16 @@ public void indexConcepts(String userName)
             if (search != null) {
                 String searchString = fieldMap.get(search.fieldName());
                 if (searchString != null) {
```
everyhting insde this if should be in a method that's being called.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -572,6 +565,16 @@ public void indexConcepts(String userName)
+            wildCardQuery.setRewriteMethod(MultiTermQuery.SCORING_BOOLEAN_REWRITE);
+            queryBuilder.add(wildCardQuery, BooleanClause.Occur.SHOULD);
+        }
+    }
```
let's document this method

jdamerow commented:
ok to test

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow Addressed review comments.

mkarthik90 commented:
Please test the search changes on UI. If the searching works as expected I will make changes to integration tests.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -536,40 +538,49 @@ public void indexConcepts(String userName)
+            BooleanQuery.Builder tokenizedQueryBuilder = new BooleanQuery.Builder();
+            buildTokenizedOrWildCardQuery(luceneFieldAnnotation, searchString, tokenizedQueryBuilder);
+            builder.add(tokenizedQueryBuilder.build(), occur);
+        }
```
this block can be merged with the previous if block. just the last line is different.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -536,40 +538,49 @@ public void indexConcepts(String userName)
+            BooleanQuery.Builder shortWordSearchQueryBuilder = new BooleanQuery.Builder();
+            shortWordSearchQueryBuilder
+                    .add(new PhraseQuery(luceneFieldAnnotation.lucenefieldName() + LuceneFieldNames.UNTOKENIZED_SUFFIX,
+                            searchString), Occur.SHOULD);
```
this too needs to use a wild card query, doesn't it?

mkarthik90 replied:
I have added the wild card query in the tokenized field. If I include the wildcard query in the untokenized short term field, the short term search doesn't work.

jdamerow commented:
retest this please

jdamerow commented:
but the functionality seems to work fine now

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow Fixed review comments and fixed integration tests.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -537,14 +530,80 @@ public void indexConcepts(String userName)
+                builder.add(new BooleanClause(
+                        new TermQuery(new Term(luceneFieldAnnotation.lucenefieldName(), searchString)), occur));
+            }
+        }
```
loose the 'else if' block and instead do a:
```
if is tokenized:
   build tokenized query (move that into the if statement, since you don't need it in the else block)
   if short term:
      build OR phrase
      assign root builder to tokenized builder
   end if short term
   add tokenized builder to builder
else:
  etc.
```
does this make sense?

jdamerow replied:
makes your if statement more concise and reduces unnecessary code

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -582,6 +641,10 @@ private void reloadReader() throws IOException {
         return wordnetIds;
     }
 
+    private boolean containsWildCardCharacters(String searchString) {
```
is that used anywhere?

jdamerow commented:
have you fixed the integration tests yet?

mkarthik90 commented:
@jdamerow  Addressed review comments. Fixed integration tests as well.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java:
```diff
@@ -537,14 +530,77 @@ public void indexConcepts(String userName)
+            QueryBuilder qBuild, BooleanQuery.Builder builder, LuceneField luceneFieldAnnotation, String searchString) {
+
+
+
```
loose all those empty lines

jdamerow commented:
ok to test

jdamerow commented:
run integration tests please

mkarthik90 commented:
@jdamerow  Removed empty lines.
